### ISSUE TITLE
Trouble with Lst - SetItem(index, value)

### ISSUE THREAD

**Issue by timminata** [louthy/language-ext#272]
Perhaps I'm missing something here but I am trying to update something in a lst at a specified index. It works for index 0 and 1 but after that things get a little strange. Here is a code snippet:

Lst<int> lint = new Lst<int>();
lint = lint.Insert(0, 0).Insert(1, 1).Insert(2, 2);
lint = lint.SetItem(2, 500);

After running this the first item is 500, then 1 then 2 whereas I was expecting 0, 1, 500.
Please advise! Also it only gets worse at higher indexes

**Comment by timminata** [comment]
Hi Paul, thanks for the quick response. This is on the latest stable version from nuget (LanguageExt.Core 2.1.21). Haven't tried building from source or anything like that as of yet. In the example it should have been Lst<int> obviously. I have also tried with other data types with the same result, however.

**Comment by louthy** [comment]
@timminata Ok, the bug I referred to was a while back, so this is new.  I will dig into it.  Although probably earliest will be Monday I'm afraid.

**Comment by louthy** [comment]
@timminata 

I just ran this code and it worked.  Do you have another example that fails?
```c#
        Lst<int> lint = new Lst<int>();
        lint = lint.Insert(0, 0).Insert(1, 1).Insert(2, 2);

        Debug.Assert(lint[0] == 0);
        Debug.Assert(lint[1] == 1);
        Debug.Assert(lint[2] == 2);

        lint = lint.SetItem(2, 500);

        Debug.Assert(lint[0] == 0);
        Debug.Assert(lint[1] == 1);
        Debug.Assert(lint[2] == 500);
```

**Comment by louthy** [comment]
@timminata Never mind, I have just seen it's the enumeration not direct access.  Leave it with me.

**Comment by timminata** [comment]
Yeah that example doesn't work for me, fails on Debug.Assert(lint[0] == 0). Thanks for taking a look

**Comment by louthy** [comment]
I've found the issue.  It's just mixed up with a few other changes that I need to be a bit careful of.  As a temporary fix, avoid `SetItem`.  You can use this below until I can deploy:
```c#
    lint = lint.RemoveAt(2).Insert(2, 500);
```

**Comment by timminata** [comment]
Cool, I will use that in the interim. Thanks.

**Comment by timminata** [comment]
Unfortunately also having problems with that solution. This code fails:

```
           Lst<int> lint = new Lst<int>();
           lint = lint.Insert(0, 0).Insert(1, 1).Insert(2, 2).Insert(3,3);

            Debug.Assert(lint[0] == 0);
            Debug.Assert(lint[1] == 1);
            Debug.Assert(lint[2] == 2);
            Debug.Assert(lint[3] == 3);

            lint = lint.RemoveAt(2).Insert(2, 500);

            Debug.Assert(lint[0] == 0);
            Debug.Assert(lint[1] == 1);
            Debug.Assert(lint[2] == 500);
            Debug.Assert(lint[3] == 3);
```

Really strange

**Comment by louthy** [comment]
This is now fixed on v2.1.22.  Sorry for the delay!

**Comment by timminata** [comment]
Thanks!
